### PR TITLE
Feature/digital input

### PR DESCRIPTION
**Pull request by LandryNorris**
Add digital input to the sample

### CODE REVIEW

**Comment by LandryNorris** [comment]
The Arduino SDK has a mechanism where calling digitalWrite will run a few checks to ensure the environment is set up properly (including setting the pin to output if it's currently set to input), and vice-versa for digitalRead. This adds significant time to the digitalRead/Write calls (this is part of why they take nearly a millisecond, while bit-banging the registers takes about a microsecond). We could implement something similar here to improve ease of use, but would also take some performance hits.

File: packages/expansion-hub/src/ExpansionHub.ts [hunk 1]
```diff
@@ -50,8 +50,21 @@ export interface ExpansionHub extends RevHub {
     // DIO
     setDigitalSingleOutput(dioPin: number, value?: boolean): Promise<void>;
     setDigitalAllOutputs(bitPackedField: number): Promise<void>;
+
+    /**
+     * Set a digital pin as input or output.
+     * @param dioPin
```
Comment (NoahAndrews) [inline packages/expansion-hub/src/ExpansionHub.ts:56]: Let's rename `dioPin` to `digitalChannel` for consistency.

File: packages/expansion-hub/src/ExpansionHub.ts [hunk 2]
```diff
@@ -50,8 +50,21 @@ export interface ExpansionHub extends RevHub {
...
     getDigitalDirection(dioPin: number): Promise<DIODirection>;
+
+    /**
+     * Read the state of a given digital pin.
+     * Throws an error if the pin is not configured for input
+     *
+     * @param dioPin
+     */
     getDigitalSingleInput(dioPin: number): Promise<boolean>;
     getDigitalAllInputs(): Promise<number>;
```
Comment (NoahAndrews) [inline packages/expansion-hub/src/ExpansionHub.ts:69]: This needs documentation.

File: packages/expansion-hub/src/ExpansionHub.ts [hunk 3]
```diff
@@ -50,8 +50,21 @@ export interface ExpansionHub extends RevHub {
...
     setDigitalDirection(dioPin: number, direction: DIODirection): Promise<void>;
     getDigitalDirection(dioPin: number): Promise<DIODirection>;
+
+    /**
+     * Read the state of a given digital pin.
+     * Throws an error if the pin is not configured for input
+     *
+     * @param dioPin
+     */
     getDigitalSingleInput(dioPin: number): Promise<boolean>;
```
Comment (NoahAndrews) [inline packages/expansion-hub/src/ExpansionHub.ts:68]: I think I'd like this to return a newly-created `DigitalState` enum:

```typescript
enum DigitalState {
    HIGH,
    LOW
}
```

Controlling a digital output should use the same enum, so let's just add support for digital outputs to this same PR.

File: packages/sample/src/main.ts [hunk 4]
```diff
@@ -1,11 +1,13 @@
 import {Command} from "commander";
 import {ExpansionHub, getConnectedExpansionHubs} from "@rev-robotics/expansion-hub";
 import {RevHub} from "@rev-robotics/expansion-hub";
+import { DIODirection } from "@rev-robotics/rhsplib";
 
 const program = new Command();
 
 program.version('1.0.0')
     .option('-l --list', 'List connected devices')
+    .option('-b --button', 'Run button press demo')
```
Comment (NoahAndrews) [inline packages/sample/src/main.ts:10]: We should be exposing primitives via the CLI, not demos.

**Review by NoahAndrews — CHANGES_REQUESTED** [review]

File: packages/sample/src/main.ts (hunk 4 above)
Comment (LandryNorris) [reply packages/sample/src/main.ts]: Changed it to use commands

File: packages/expansion-hub/src/ExpansionHub.ts (hunk 1 above)
Comment (LandryNorris) [reply packages/expansion-hub/src/ExpansionHub.ts:65]: Renamed

**Comment by NoahAndrews** [comment]
This is difficult to review in its current form, since right now GH displays it as including the prettier PR for some reason

File: packages/expansion-hub/src/ExpansionHub.ts [hunk 5]
```diff
@@ -57,11 +58,28 @@ export interface ExpansionHub extends RevHub {
...
-    setDigitalSingleOutput(dioPin: number, value?: boolean): Promise<void>;
+    setDigitalSingleOutput(digitalChannel: number, value: DigitalState): Promise<void>;
     setDigitalAllOutputs(bitPackedField: number): Promise<void>;
+
+    /**
+     * Set a digital pin as input or output.
+     * @param dioPin
+     * @param direction
+     */
     setDigitalDirection(dioPin: number, direction: DIODirection): Promise<void>;
```
Comment (NoahAndrews) [inline packages/expansion-hub/src/ExpansionHub.ts:69]: `DIODirection` should be `DioDirection` or maybe `DigitalChannelDirection`

File: packages/expansion-hub/src/ExpansionHub.ts [hunk 6]
```diff
@@ -57,11 +58,28 @@ export interface ExpansionHub extends RevHub {
...
     getDigitalDirection(dioPin: number): Promise<DIODirection>;
-    getDigitalSingleInput(dioPin: number): Promise<boolean>;
+
+    /**
+     * Read the state of a given digital pin.
+     * Throws an error if the pin is not configured for input
+     *
+     * @param dioPin
+     */
+    getDigitalSingleInput(dioPin: number): Promise<DigitalState>;
```
Comment (NoahAndrews) [inline packages/expansion-hub/src/ExpansionHub.ts:78]: `dioPin` needs to be `digitalChannel`

Comment (NoahAndrews) [reply packages/expansion-hub/src/ExpansionHub.ts:78]: I think a better name would be `getDigitalInput()` (`getDigitalAllInputs() would become `getAllDigitalInputs()`). Outputs would follow the same scheme.

**Review by NoahAndrews — CHANGES_REQUESTED** [review]

File: packages/expansion-hub/src/ExpansionHub.ts (hunk 5 above)
Comment (LandryNorris) [reply packages/expansion-hub/src/ExpansionHub.ts]: Changing DIODirection would be a breaking change in rhsplib. I'd prefer to create DigitalChannelDirection in expansion-hub and convert internally.

File: packages/expansion-hub/src/ExpansionHub.ts (hunk 6 above)
Comment (LandryNorris) [reply packages/expansion-hub/src/ExpansionHub.ts]: Renamed the methods.

File: packages/expansion-hub/src/ExpansionHub.ts (hunk 5 above)
Comment (NoahAndrews) [reply packages/expansion-hub/src/ExpansionHub.ts]: Fine with me

File: packages/expansion-hub/src/ExpansionHub.ts [hunk 7]
```diff
@@ -57,12 +58,37 @@ export interface ExpansionHub extends RevHub {
...
-    getDigitalDirection(dioPin: number): Promise<DIODirection>;
-    getDigitalSingleInput(dioPin: number): Promise<boolean>;
-    getDigitalAllInputs(): Promise<number>;
+    /**
+     * Set the output state of a digital pin.
+     * @param digitalChannel
+     * @param value whether the pin should be High or Low
+     */
+    setDigitalOutput(digitalChannel: number, value: DigitalState): Promise<void>;
+    setAllDigitalOutputs(bitPackedField: number): Promise<void>;
```
Comment (NoahAndrews) [inline packages/expansion-hub/src/ExpansionHub.ts:67]: Needs a doc comment

File: packages/sample/src/main.ts [hunk 8]
```diff
@@ -20,4 +22,26 @@ program
...
     });
 
+let digitalCommand = program.command("digital");
+
+digitalCommand
+    .command("write <channel> <state>")
+    .description("write digital pin")
+    .action(async (channel, state) => {
+        let channelNumber = Number(channel);
+        let stateValue = state == "high" ? DigitalState.High : DigitalState.Low;
```
Comment (NoahAndrews) [inline packages/sample/src/main.ts:32]: We need to document what values to pass in for `<state>`, and we should reject anything besides `high`, `low`, `1`, or `0`, and process `1` correctly.

**Review by NoahAndrews — CHANGES_REQUESTED** [review]

File: packages/expansion-hub/src/ExpansionHub.ts (hunk 7 above)
Comment (LandryNorris) [reply packages/expansion-hub/src/ExpansionHub.ts:67]: Do we know how this method reacts if any pins are set to input? Will it skip the input pins, or will it throw a nack?

File: packages/sample/src/main.ts (hunk 8 above)
Comment (LandryNorris) [reply packages/sample/src/main.ts]: Added list of valid entries to help and added code to reject anything other than high, low, 0, or 1.

**Comment by NoahAndrews** [comment]
I'll re-review after the merge conflicts are fixed.

File: packages/core/src/digital-state.ts
Comment (NoahAndrews) [inline packages/core/src/digital-state.ts:1]: File should be named `DigitalState.ts`

File: packages/core/src/digital-state.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
+export class DigitalState {
+    static readonly High = new DigitalState(true);
+    static readonly Low = new DigitalState(false);
```
Comment (NoahAndrews) [inline packages/core/src/digital-state.ts:3]: These should be in ALL_CAPS_SNAKE_CASE

**Review by NoahAndrews — CHANGES_REQUESTED** [review]

**Review by NoahAndrews — APPROVED** [review]

**Pull request merged**